### PR TITLE
Making Imputation generic

### PR DESCRIPTION
**Pull request by thomashenson**
There is still more work to be done on Imputation to make it truely generic, however we have completed the work as outlined in the story and will look to continue work inside a seperate story in the next sprint.

### CODE REVIEW

**Comment by Piwington** [comment]
Noticed We Missed Out On Adding A Change. Tom, Mike And Myself Are Resolving Now But May Impact Reviewing.

File: calculate_movement_method.py [hunk 1]
```diff
@@ -36,6 +39,13 @@ def lambda_handler(event, context):
     final_output = {}
 
     try:
+        # TESTING
```
Comment (KrisRogos) [inline calculate_movement_method.py]: What are you testing? Why is this still here going into a PR?

File: calculate_movement_method.py [hunk 2]
```diff
@@ -47,24 +57,28 @@ def lambda_handler(event, context):
...
 
             result_list = []
+
             # .Length is used so the correct amount of iterations for the loop.
             for i in range(0, len(sorted_current)):
 
                 # This check is too prevent the DivdebyZeroError.
                 if previous_list[i] != 0:
-                    number = (current_list[i] - previous_list[i]) / previous_list[i]
+                    # TESTING
```
Comment (KrisRogos) [inline calculate_movement_method.py]: Is this still being tested?

File: calculate_movement_wrangler.py [hunk 3]
```diff
@@ -189,14 +193,19 @@ def lambda_handler(event, context):
...
+                    merged_data,
+                    period, time,
+                    reference, segmentation,
+                    stored_segmentation,
+                    current_time,
+                    previous_time,
+                    current_segmentation,
+                    previous_segmentation)
 
             logger.info("Successfully completed strata mismatch detection")
```
Comment (KrisRogos) [inline calculate_movement_wrangler.py]: Should this be indented to be inside the `if` to only fire if strata_mismatch_detector was actually run?

File: imputation_functions.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    :return: List of column names with desired prefix - Type: List
+    """
+    new_columns = []
+    for column in columns:
+        new_value = "%s%s%s" % (prefix, column, suffix)
+        new_columns.append(new_value)
+
+    new_columns = new_columns + additional
+
+    return new_columns
```
Comment (KrisRogos) [inline imputation_functions.py]: Doesn't this function also exist in the apply_factors_wrangler? Should be only in one place if possible. The comment on the suffix being used as a string but described as a list still applies.

File: iqrs_method.py [hunk 5]
```diff
@@ -89,18 +89,32 @@ def lambda_handler(event, context):
...
-def calc_iqrs(input_table, move_cols, iqrs_cols):
-    distinct_strata_region = input_table[['region', 'strata']].drop_duplicates()
+def calc_iqrs(input_table, move_cols, iqrs_cols, distinct_values):
+    """
+    Calculate IQRS.
+    :param input_table: Input data.
+    :param move_cols: Movement column names.
+    :param iqrs_cols: IQRS column names.
+    :param distinct_values: Array of column names to derive distinct values from
+                            and store in table.
```
Comment (KrisRogos) [inline iqrs_method.py]: No types

File: iqrs_wrangler.py [hunk 6]
```diff
@@ -70,13 +72,20 @@ def lambda_handler(event, context):
...
         wrangled_data = lambda_client.invoke(
             FunctionName=method_name,
-            Payload=json.dumps(data_json)
+            Payload=payload
         )
 
         json_response = wrangled_data.get('Payload').read().decode("UTF-8")
-
+        import pandas as pd
+        json_response   # HERERERERERERE
```
Comment (KrisRogos) [inline iqrs_wrangler.py]: ?

File: apply_factors_wrangler.py [hunk 7]
```diff
@@ -244,3 +253,22 @@ def lambda_handler(event, context):
...
+    Produces columns with a prefix, based on standard columns.
+    :param prefix: String to be prepended to column name - Type: String
+    :param columns: List of columns - Type: List
+    :param suffix: Any additonal columns to be added on - Type: List
+
+    :return: List of column names with desired prefix - Type: List
+    """
+    new_columns = []
+    for column in columns:
+        new_value = "%s%s%s" % (prefix, column, suffix)
```
[the remark this answers is not in the document]
Comment (thomashenson) [reply apply_factors_wrangler.py]: Does need to be updated, will sort this now.

[the remark this answers is not in the document]
Comment (thomashenson) [reply apply_factors_wrangler.py]: Actually this should now be removed from atyps as is now placed in the imp functions file @Piwington

[the remark this answers is not in the document]
Comment (KrisRogos) [reply apply_factors_wrangler.py]: Yes, it is, thanks

File: imputation_functions.py (hunk 4 above)
Comment (KrisRogos) [reply imputation_functions.py]: Only one copy with the correct comment left, resolved

File: calculate_movement_method.py (hunk 1 above)
Comment (thomashenson) [reply calculate_movement_method.py]: Resolved.

File: calculate_movement_wrangler.py (hunk 3 above)
Comment (thomashenson) [reply calculate_movement_wrangler.py]: Fixed.

File: apply_factors_wrangler.py [hunk 8]
```diff
@@ -150,16 +143,34 @@ def lambda_handler(event, context):
...
 
-        # Filter Out The Data Columns From The Temporary Calculated Columns.
-        filtered_data = final_imputed[['Q601_asphalting_sand', 'Q602_building_soft_sand',
-                                       'Q603_concreting_sand', 'Q604_bituminous_gravel',
-                                       'Q605_concreting_gravel', 'Q606_other_gravel',
-                                       'Q607_constructional_fill', 'Q608_total', 'county',
-                                       'county_name', 'enterprise_ref', 'gor_code',
-                                       'land_or_marine', 'name', 'period', 'region',
-                                       'responder_id', 'strata']]
+        # See Mike
```
Comment (kingmushroom) [inline apply_factors_wrangler.py]: Not sure how i feel about this....

File: tests/test_apply_factors.py [hunk 9]
```diff
@@ -331,8 +381,10 @@ def test_wrangles_incomplete_data(self):
             message = file.read()
         with open("tests/fixtures/test_data.json", "r") as file:
             prevfile = pd.DataFrame(json.loads(file.read()))
+            print(type(prevfile))
```
Comment (kingmushroom) [inline tests/test_apply_factors.py]: prints?

File: apply_factors_wrangler.py [hunk 10]
```diff
@@ -146,7 +146,12 @@ def lambda_handler(event, context):
 
         logger.info("Successfully joined imputed data with responder data")
 
-        # See Mike
+        # Create A List Of Movement Columns Plus (
+        #   A List Of Means Columns Plus (
+        #       A List Of Factor Columns Plus (
+        #           A List Of Movement Sum Columns Plus (
+        #               A List Of Movement Count Columns))))
+        # See Mike For Why This Way And Not Variables Appended Together.
```
Comment (kingmushroom) [inline apply_factors_wrangler.py]: Still not sure how i feel about this. The explanation works nicely though.

File: tests/test_means.py [hunk 11]
```diff
@@ -159,7 +178,8 @@ def test_wrangler_key_error(self):
 
     def test_method_key_error(self):
         # pass none value to trigger key index error
-        response = calculate_means_method.lambda_handler(None, context_object)
+        response = calculate_means_method.lambda_handler({"mike": "mike"}, context_object)
+        print(response)
```
Comment (kingmushroom) [inline tests/test_means.py]: print

**Comment by kingmushroom** [comment]
Looks ok from my end.

File: iqrs_method.py [hunk 12]
```diff
@@ -89,18 +79,32 @@ def lambda_handler(event, context):
...
+            iqr_filter += " & "
+        iqr_filter += "(input_table[\"%s\"] == row[%s])"\
+            % (value, distinct_values.index(value))
+
     for row in distinct_strata_region.values:
-        iqr_filter = (input_table["region"] == row[0]) & (input_table["strata"] == row[1])  # noqa: E501
-        filtered_iqr = input_table[iqr_filter]
-        # Pass the question number and region and strata groupping to the
-        # iqr_sum function.
+        filtered_iqr = input_table[pd.eval(iqr_filter)]
```
Comment (glanvl) [inline iqrs_method.py]: I was under the impression we didn't want to be using eval? Ian had asked Tom to take eval out completely so he may want us to here.

File: iqrs_method.py [hunk 13]
```diff
@@ -89,18 +79,32 @@ def lambda_handler(event, context):
...
+        # Pass the question number and region and strata grouping to the iqr_sum function.
         for i in range(0, len(iqrs_cols)):
             val_one = iqr_sum(filtered_iqr, move_cols[i])
             input_table[iqrs_cols[i]] = np.where(
-                ((input_table["region"] == row[0]) & (input_table["strata"] == row[1])), val_one,  # noqa: E501
-                input_table[iqrs_cols[i]]
+                (pd.eval(iqr_filter)), val_one, input_table[iqrs_cols[i]]
```
Comment (glanvl) [inline iqrs_method.py]: Eval used again see previous comment

**Pull request merged**